File: VB/Q211932/MyDateEdit.vb
Imports DevExpress.XtraEditors
Imports DevExpress.XtraEditors.Repository
Imports DevExpress.XtraEditors.Registrator
Imports DevExpress.XtraEditors.ViewInfo
Imports DevExpress.XtraEditors.Drawing
Imports System.ComponentModel
Imports DevExpress.XtraEditors.Popup
Imports DevExpress.XtraEditors.Controls
Imports System
Imports System.Windows.Forms

Namespace DXSample
	Public Class MyDateEdit
		Inherits DateEdit

		Shared Sub New()
			RepositoryItemMyDateEdit.RegisterMyDateEdit()
		End Sub

		Public Sub New()
			MyBase.New()
		End Sub

		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return RepositoryItemMyDateEdit.MyDateEditName
			End Get
		End Property

		<DesignerSerializationVisibility(DesignerSerializationVisibility.Content)>
		Public Shadows ReadOnly Property Properties() As RepositoryItemMyDateEdit
			Get
				Return CType(MyBase.Properties, RepositoryItemMyDateEdit)
			End Get
		End Property

		Protected Overrides Function CreatePopupForm() As PopupBaseForm
			If Properties.IsVistaDisplayModeInternal() Then
				Return New MyVistaPopupDateEditForm(Me)
			End If
			Return MyBase.CreatePopupForm()
		End Function

		Protected Overrides Sub OnKeyDown(ByVal e As KeyEventArgs)
			If e.KeyCode = Keys.Delete Then
				EditValue = Nothing
				e.Handled = True
			Else
				MyBase.OnKeyDown(e)
			End If
		End Sub
	End Class

	<UserRepositoryItem("RegisterMyDateEdit")>
	Public Class RepositoryItemMyDateEdit
		Inherits RepositoryItemDateEdit

		Shared Sub New()
			RegisterMyDateEdit()
		End Sub

		Public Sub New()
			MyBase.New()
		End Sub

		Friend Const MyDateEditName As String = "MyDateEdit"

		Public Overrides ReadOnly Property EditorTypeName() As String
			Get
				Return MyDateEditName
			End Get
		End Property

		Public Shared Sub RegisterMyDateEdit()
			EditorRegistrationInfo.Default.Editors.Add(New EditorClassInfo(MyDateEditName, GetType(MyDateEdit), GetType(RepositoryItemMyDateEdit), GetType(DateEditViewInfo), New ButtonEditPainter(), True))
		End Sub

		Friend Function IsVistaDisplayModeInternal() As Boolean
			Return IsVistaDisplayMode()
		End Function
	End Class

	Public Class MyVistaPopupDateEditForm
		Inherits VistaPopupDateEditForm

		Public Sub New(ByVal ownerEdit As DateEdit)
			MyBase.New(ownerEdit)
		End Sub

		Protected Overrides Function CreateCalendar() As CalendarControl
			Dim result As New MyVistaDateEditCalendar()

			Return result
		End Function

	End Class

	Public Class MyVistaDateEditCalendar
		Inherits CalendarControl

		Public Sub New()

		End Sub
		Protected Overrides Sub OnKeyDown(ByVal e As KeyEventArgs)
			If e.KeyCode = Keys.Delete Then
				OnDateTimeCommit(Nothing)
			 e.Handled = True
			 Return
			End If
			MyBase.OnKeyDown(e)
		End Sub
	End Class
End Namespace

File: CS/Q211932/MyDateEdit.cs
using DevExpress.XtraEditors;
using DevExpress.XtraEditors.Repository;
using DevExpress.XtraEditors.Registrator;
using DevExpress.XtraEditors.ViewInfo;
using DevExpress.XtraEditors.Drawing;
using System.ComponentModel;
using DevExpress.XtraEditors.Popup;
using DevExpress.XtraEditors.Controls;
using System;
using System.Windows.Forms;

namespace DXSample {
    public class MyDateEdit : DateEdit {
        static MyDateEdit() { RepositoryItemMyDateEdit.RegisterMyDateEdit(); }

        public MyDateEdit() : base() { }

        public override string EditorTypeName { get { return RepositoryItemMyDateEdit.MyDateEditName; } }

        [DesignerSerializationVisibility(DesignerSerializationVisibility.Content)]
        public new RepositoryItemMyDateEdit Properties { 
            get { 
                return (RepositoryItemMyDateEdit)base.Properties; 
            } 
        }

        protected override PopupBaseForm CreatePopupForm() {
            if (Properties.IsVistaDisplayModeInternal()) return new MyVistaPopupDateEditForm(this);
            return base.CreatePopupForm();
        }

        protected override void OnKeyDown(KeyEventArgs e) {
            if (e.KeyCode == Keys.Delete) {
                EditValue = null;
                e.Handled = true;
            } else base.OnKeyDown(e);
        }
    }

    [UserRepositoryItem("RegisterMyDateEdit")]
    public class RepositoryItemMyDateEdit : RepositoryItemDateEdit {
        static RepositoryItemMyDateEdit() { RegisterMyDateEdit(); }

        public RepositoryItemMyDateEdit() : base() { }
        
        internal const string MyDateEditName = "MyDateEdit";
        
        public override string EditorTypeName { get { return MyDateEditName; } }

        public static void RegisterMyDateEdit() {
            EditorRegistrationInfo.Default.Editors.Add(new EditorClassInfo(MyDateEditName, typeof(MyDateEdit),
                typeof(RepositoryItemMyDateEdit), typeof(DateEditViewInfo), new ButtonEditPainter(), true));
        }

        internal bool IsVistaDisplayModeInternal() { return IsVistaDisplayMode(); }
    }

    public class MyVistaPopupDateEditForm : VistaPopupDateEditForm {
        public MyVistaPopupDateEditForm(DateEdit ownerEdit) : base(ownerEdit) { }

        protected override DateEditCalendar CreateCalendar() {
            MyVistaDateEditCalendar result = new MyVistaDateEditCalendar(OwnerEdit.Properties,
                OwnerEdit.EditValue);
            result.OkClick += new EventHandler(OnOkClick);
            return result;
        }
    }

    public class MyVistaDateEditCalendar : VistaDateEditCalendar {
        public MyVistaDateEditCalendar(RepositoryItemDateEdit item, object date) : base(item, date) { }

        protected override void OnKeyDownCore(KeyEventArgs e) {
            if (e.KeyCode == Keys.Delete) {
                OnDateTimeCommit(null, true);
                e.Handled = true;
                return;
            }
            base.OnKeyDownCore(e);
        }
    }
}